[WIP] [shelly] Add Shelly Blu Wall Switch 4 and Shelly BLU RC Button 4

UdoKrie opened the pull request:
This PR implements support for the Shelly devices based on BLU Button 4, known as Shelly Blu Wall Switch 4 and Shelly BLU RC Button 4.

Mainly the transportation scheme from a device used as BLU gateway to OH shelly binding is adapted:
- JSON serialization class Shelly2NotifyEventMessage
  - member Button changed to Buttons, instead of a single byte now a byte array is transported where the index ([0..3]) within the array represents the button ID [1..4]
  - the BLU gateway script oh-blu-scanner.js is extended so that it first collects all 0x3A events in an array befor adding to the JSON serializer
  - BLU Button1 is now simply a subset of BLU Button 4
- BLU Button 4 supports a new state `HOLD`, an according event type is added, which may be used for interessting use cases like continuous dimming (light) or sliding (roller shutter)

Regarding the device profile:
- I kindly ask the reviewers to check, if their might be an approach to cover these new devices mapping to existing code, that channel discovery took me a while to figure out and might be coverable with existing means.

For testing find here a [pre-build](https://github.com/UdoKrie/openhab-dev/blob/main/shelly/OH5.0.x/org.openhab.binding.shelly-5.0.0-SNAPSHOT.jar) jar

markus7017 commented:
great stuff, extended support for the BLU buttons was on my list. I'll come back to the PR, maybe by end of the week (traveling)

You might involve the community for testing, just post here: https://community.openhab.org/t/shelly-binding/56862/4571?u=markus7017

markus7017 commented:
I saw that there are a lot of changes in shelly.properties, because of a trailing blank.
Could you please extract those into a separate PR, because it's not related to the functional changes of this PR and als streamlines the changes here. The new PR can be merged independently.

markus7017 commented:
@UdoKrie I tested with BLU RC Button 4 and saw the following issues
- Button 4 has model id SBBT-EU3870", you code only handles the Wall Switch (model id SBBT-EU5027)
- the binding reports an unknown BLU device when pressing the button (see below)

When looking to
```
    public static void addBluThing(String gateway, Shelly2NotifyEvent e, @Nullable ShellyThingTable thingTable) {
        String model = substringBefore(getString(e.data.name), "-").toUpperCase();
        String mac = e.data.addr.replaceAll(":", "");
        String ttype = "";
        LOGGER.debug("{}: Create thing for new BLU device {}: {} / {}", gateway, e.data.name, model, mac);
        ThingTypeUID tuid;
        switch (model) {
            case "SBBT":
            case SHELLYDT_BLUBUTTON:
                ttype = THING_TYPE_SHELLYBLUBUTTON_STR;
                tuid = THING_TYPE_SHELLYBLUBUTTON;
                break;
            case SHELLYDT_BLUWALLSWITCH4:
                ttype = THING_TYPE_SHELLYBLUWALLSWITCH4_STR;
                tuid = THING_TYPE_SHELLYBLUWALLSWITCH4;
                break;
```
this can't work from my point of view, because 
  model = substringBefore(getString(e.data.name), "-").toUpperCase(); 
costs off the model id so the switch() compares "SBBT" with "SBBT-EU3870", which fails.

I think first decoding the class and then the model id could work
```
   public static void addBluThing(String gateway, Shelly2NotifyEvent e, @Nullable ShellyThingTable thingTable) {
        String model = substringBefore(getString(e.data.name), "-").toUpperCase();
        String subModel = substringAfter(getString(e.data.name), "-").toUpperCase();
        String mac = e.data.addr.replaceAll(":", "");
        String ttype = "";
        LOGGER.debug("{}: Create thing for new BLU device {}: {} / {}", gateway, e.data.name, model, mac);
        ThingTypeUID tuid;
        switch (model) {
            case "SBBT":
                switch (getString(e.data.name)) {
                    case SHELLYDT_BLUBUTTON:
                    default:
                        ttype = THING_TYPE_SHELLYBLUBUTTON_STR;
                        tuid = THING_TYPE_SHELLYBLUBUTTON;
                        break;
                    case SHELLYDT_BLURCBUTTON4:
                        ttype = THING_TYPE_SHELLYBLURCBUTTON4_STR;
                        tuid = THING_TYPE_SHELLYBLURCBUTTON4;
                        break;
                    case SHELLYDT_BLUWALLSWITCH4:
                        ttype = THING_TYPE_SHELLYBLUWALLSWITCH4_STR;
                        tuid = THING_TYPE_SHELLYBLUWALLSWITCH4;
                        break;
                }
                break;
            case SHELLYDT_BLUDW:
```

same in buildBluServiceName()

Are you sure you pushed the lastest code?
Did you tried only Wall Switch 4 and assumed that RC Button 4 also works?

Please add the additional handling and a separate thing type for RC Button 4

UdoKrie commented:
@markus7017 
Thanks for feedback.
Regarding `addBluThing()` your suggestions makes sense, I'll adapt accordingly.
Regarding RC4 , I don't own one. The Shelly BLU APi documention state `SBBT-004CUS` as devide name, can you confirm that your device actually provides this string ? Because my wall switch send SBBT-EU3870 despite SBBT-004CEU is documented.

Reagrding isMultiButton: I'm not sure an additional member isMultiButton is really required or wether another member like isIX could be reused here. Usage of those flags is not necessarily easy to track. Do you have an advice on this?

markus7017 commented:
I have both: Wall Switch 4 and RC Button 1
Correct, I also noticed the mismatch between documented device ids and those reported by the device. I could verify those for the 2 devices mentioned above. We could add the US ones and may need to change them later.

Let's work on my comments and then could do some testing with both devices (I have a 3rd, which is not yet launched).

UdoKrie commented:
@markus7017 
Hi Markus
I will work on this commit soon, probably already tomorrow evening. Today is occupied with bringing data of my new solar battery into openHAB.
br,
Udo

UdoKrie commented:
@markus7017 
I added the RC Button 4 thing, however, couldn't test it, hope, I didn't miss anything.
Also I implemented your suggestions so far.
This commit is not yet rebased, because my local rebased branch failed when testing. I assume it's related to the changes in shellyThingCreator and  my extension in the contructor ShellyBluApi(). Maybe you can give me a hint on that.
br,
Udo

UdoKrie commented:
> ....
> Because my wall switch send SBBT-EU3870 despite SBBT-004CEU is documented.
> ...

Looks like there is somewhere a flaw when the device string is constructed.
In my case the full device name should look like `SBBT-004CEU-hhhh`, where _hhhh_ are the last four hex digits of the mac address, in my case `SBBT-004CEU-3870`. I will check how the name is propagated from BLE.Scanner.start() into ShellyBluApi().
`SBBT-EU3870`, what I've seen during debugging, is definitely incorrect.

jlaur commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api2/ShellyBluApi.java:
```diff
@@ -60,33 +61,45 @@ public class ShellyBluApi extends Shelly2ApiRpc {
-            "2", SHELLY_BTNEVENT_2SHORTPUSH, //
-            "3", SHELLY_BTNEVENT_3SHORTPUSH, //
-            "4", SHELLY_BTNEVENT_LONGPUSH);
+    private static final Map<String, String> MAP_INPUT_EVENT_TYPE = Map.ofEntries( //
```
It's the only way to provide more than 10 key/value pairs. See https://stackoverflow.com/questions/46601959/what-is-the-use-of-map-ofentries-instead-of-map-of

lsiepel commented:
@UdoKrie please rebase / solve the conflicts

markus7017 commented:
@UdoKrie Please rebase and works on my comments.
I'll come back to this PR after 5.0 has been released next week. Fixes and the Gen4 have prior at the moment even I like this feature :-)

wborn commented:
Let's close this now this is part of https://github.com/openhab/openhab-addons/pull/19097
